rfcs/134-write-behavior-consistency.md

```diff
@@ -0,0 +1,234 @@
+- [Make write Idempotent](https://github.com/beyondstorage/specs/issues/96)
+- [copy/move doesn't need to support dst path that has an existing path](https://github.com/beyondstorage/go-integration-test/issues/29)
+
+## Background
+
+Behavior consistency is extremely critical to us, and write related operations is the most important part.
+
+Write related operations including following operations:
+
+- `Write`
+- `CompleteMultipart`
+- `CombineBlock`
+- `CommitAppend`
+- `Copy`
+- `Move`
+
+All of them will create a new object in our storage service. But our services could have different behavior while the destination path already exists. I will list a few representative service behaviors for analysis.
+
+### fs
+
+For `Write`:
+
+`fs` could return `EEXIST` to indicate `File exists`. Users can set open mode `Create` to create a new file here without meeting `EEXIST` error.
+
+For `Copy`:
+
+`fs` doesn't have native copy support, we could implement it via reading from src and write into dst. For the writing part, most userland tools will use `Create` open mode to ignore `EEXIST` error.
+
+So `Copy` will not return the `EEXIST` error.
+
+For `Move`:
+
+`fs` will ignore the `EEXIST` error and always replace the file. On the same filesystem, it's an atomic operation, otherwise, it should be implemented as `copy-delete`.
+
+So `Move` will not return the `EEXIST` error.
+
+### s3
+
+Most s3 alike object storage services will not return object exist errors.
+
+- For object storage services that don't have an object version, the last write will always succeed.
+- For object storage services that have object versions, every write will create a new object version.
+
+### ipfs
+
+For `Write`:
+
+`ipfs` supports `Create` like `fs`. With `Create`, `ipfs` will create the file if it does not exist, this is the only behavior, in other words, `ipfs` will continue to write to the file even if create is not specified.
+
+If writing to an existing file, IPFS will overwrite it from the beginning, and if there are unwritten parts at the end of the original file, they will remain as well. `ipfs` provides an option `Truncate` to make sure that the write is clean.
+
+```go
+func (s *Storage) write(ctx context.Context, path string, r io.Reader, size int64, opt pairStorageWrite) (n int64, err error) {
+   err = s.ipfs.FilesWrite(
+      ctx, s.getAbsPath(path), r,
+      ipfs.FilesWrite.Create(true),
+      ipfs.FilesWrite.Parents(true),
+      ipfs.FilesWrite.Truncate(true),
+   )
+   if err != nil {
+      return 0, err
+   }
+   return size, nil
+}
+```
+
+For `Copy` and `Move`:
+
+`ipfs` have native support for `Copy` and `Move`. If the dst exists, the `Copy` operation will return `FileExists`, while the `Move` operation will overwrite.
+
+### dropbox
+
+For `Write`:
+
+`dropbox` supports setting `WriteMode`. With `files.WriteModeAdd`, `Write` will not return `FileExists` too:
```

Comment:
Should  `files.WriteModeAdd` be `files.WriteModeOverwrite`? With `files.WriteModeAdd`, `Write` will succeed if the file at the target path with contents same with the contents you're trying to write, but `path/conflic/file` if not.

Reply:
The logic is borrowed from `go-service-dropbox`. Maybe we can send an issue to address this problem?

Reply:
Maybe we can update here directly as the background reference to the official doc? Actually consistency issue is not considered before and all the services should be updated if needed.

Reply:
> Maybe we can update here directly as the background reference to the official doc? Actually consistency issue is not considered before and all the services should be updated if needed.

In this GSP, I just want to give some examples to show out different behavior that services provide, and don't want to address all of them here. I think we need to track consistency issues in different services or creating a tracking issue in `go-storage`?